File: Kotlin/ExtratoService.kt
package com.kamino.banking.service

import com.fasterxml.jackson.databind.ObjectMapper
import com.kamino.banking.entity.*
import com.kamino.banking.repository.*
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.cache.annotation.Cacheable
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.client.RestTemplate
import java.math.BigDecimal
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*

@Service
class ExtratoService(
    val transacaoRepository: TransacaoFinanceiraRepository,
    val contaBancoRepository: ContaBancoRepository,
    val contaPagarRepository: ContaPagarRepository,
    val restTemplate: RestTemplate,
    val kafkaTemplate: KafkaTemplate<String, String>,
    val objectMapper: ObjectMapper
) {
    private val logger = LoggerFactory.getLogger(ExtratoService::class.java)

    @Value("\${kamino.api.url}")
    lateinit var apiUrl: String

    @Value("\${kamino.api.token}")
    lateinit var apiToken: String

    private val saldoCache = HashMap<String, BigDecimal>()

    fun consultarPendentes(companyName: String, idContaBanco: Int): List<Map<String, Any>>? {
        try {
            val url = "$apiUrl/hub/v1/financial/transactions/pendings"
            val headers = HttpHeaders()
            headers.set("Authorization", "Bearer $apiToken")
            headers.set("X-Company", companyName)
            headers.set("X-Account-Id", idContaBanco.toString())

            val response = restTemplate.exchange(url, HttpMethod.GET, HttpEntity<Any>(headers), String::class.java)

            if (response.statusCode.is2xxSuccessful) {
                return objectMapper.readValue(response.body, List::class.java) as List<Map<String, Any>>
            } else {
                logger.error("Erro ao buscar transações pendentes: ${response.statusCode}")
            }
        } catch (ex: Exception) {
            logger.error("Erro ao buscar transações pendentes", ex)
        }
        return null
    }

    @Cacheable("saldoPendente")
    fun getSaldoPendente(companyName: String, idContaBanco: Int): Map<String, Any>? {
        val cacheKey = "${companyName}_$idContaBanco"
        if (saldoCache.containsKey(cacheKey)) {
            return mapOf("balance" to saldoCache[cacheKey]!!, "cached" to true)
        }

        try {
            val url = "$apiUrl/hub/v1/bank/balance/pendings"
            val headers = HttpHeaders()
            headers.set("Authorization", "Bearer " + apiToken)
            headers.set("X-Company", companyName)

            val response = restTemplate.exchange(url, HttpMethod.GET, HttpEntity<Any>(headers), Map::class.java)
            val balance = response.body?.get("balance") as? BigDecimal

            if (balance != null) {
                saldoCache[cacheKey] = balance
            }
            return response.body as Map<String, Any>?
        } catch (ex: Exception) {
            logger.error("Erro ao buscar saldo pendente: ${ex.message}")
        }
        return null
    }

    fun atualizarContasBancarias(companyToken: String, forcar: Boolean = false): RetornoHelper {
        var reagendar = true
        try {
            var contaBancos = contaBancoRepository.findContasIntegradas()

            if (contaBancos.isEmpty()) {
                reagendar = false
                return RetornoHelper(sucesso = true)
            }

            logger.info("Atualizando ${contaBancos.size} contas bancarias")

            for (contaBanco in contaBancos) {
                sincronizarConta(contaBanco, companyToken, forcar)
            }

            return RetornoHelper(sucesso = true)
        } catch (ex: Exception) {
            agendarSincronizacao(companyToken)
            logger.error("Erro ao atualizar extrato contas bancárias", ex)
            return RetornoHelper(sucesso = false, mensagem = "Não foi possível atualizar as contas bancárias")
        } finally {
            if (reagendar) {
                agendarSincronizacao(companyToken)
            }
        }
    }

    fun agendarSincronizacao(companyToken: String) {
        try {
            val message = mapOf(
                "companyToken" to companyToken,
                "metodo" to "POST",
                "endpoint" to "api/banking/kamino/transactions/reload/hook",
                "dataHoraProgramacao" to LocalDateTime.now().plusMinutes(5).toString()
            )
            kafkaTemplate.send("broker-queue", objectMapper.writeValueAsString(message))
        } catch (ex: Exception) {
            logger.error("Erro ao agendar broker de atualização", ex)
        }
    }

    @Async
    @Transactional
    fun sincronizarConta(contaBanco: ContaBanco, companyToken: String, forcar: Boolean = false): RetornoHelper {
        logger.info("Iniciando atualização da conta bancaria: ${contaBanco.id}")

        val ultimaTransacao = transacaoRepository.findUltimaTransacao(contaBanco.id)
        
        var dataInicio = ultimaTransacao?.data ?: LocalDateTime.of(2023, 1, 1, 0, 0)
        dataInicio = dataInicio.minusDays(1)
        if (forcar) dataInicio = LocalDateTime.of(2023, 1, 1, 0, 0)

        val transacoesExternas = buscarTransacoesAPI(contaBanco.id, dataInicio, LocalDateTime.now())
        logger.info("Transações obtidas da API: ${transacoesExternas.size}")

        for (transacaoExterna in transacoesExternas) {
            val transacao = converterParaEntidade(transacaoExterna, contaBanco)
            salvarTransacao(transacao)
        }

        calcularSaldoDiario(contaBanco, dataInicio)

        return RetornoHelper(sucesso = true)
    }

    private fun buscarTransacoesAPI(idContaBanco: Int, dataInicio: LocalDateTime, dataFim: LocalDateTime): List<Map<String, Any>> {
        val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
        val url = "$apiUrl/hub/v1/financial/transactions?startDate=${dataInicio.format(formatter)}&finalDate=${dataFim.format(formatter)}&accountId=$idContaBanco"

        val headers = HttpHeaders()
        headers.set("Authorization", "Bearer $apiToken")

        try {
            val response = restTemplate.exchange(url, HttpMethod.GET, HttpEntity<Any>(headers), Map::class.java)
            return (response.body?.get("items") as? List<Map<String, Any>>) ?: emptyList()
        } catch (ex: Exception) {
            logger.error("Erro ao buscar transações da API", ex)
            return emptyList()
        }
    }

    private fun converterParaEntidade(dados: Map<String, Any>, contaBanco: ContaBanco): TransacaoFinanceira {
        val transacao = TransacaoFinanceira()
        transacao.codigoNoBanco = dados["id"] as? String
        transacao.descricao = dados["description"] as? String
        transacao.valor = BigDecimal(dados["amount"].toString())
        transacao.data = LocalDateTime.parse(dados["date"] as String)
        transacao.idContaBanco = contaBanco.id
        transacao.idTransacaoExterna = dados["externalId"]?.let { UUID.fromString(it as String) }
        return transacao
    }

    fun salvarTransacao(transacao: TransacaoFinanceira) {
        val existente = transacaoRepository.findByCodigoNoBanco(transacao.codigoNoBanco ?: "")
        
        if (existente != null) {
            existente.descricao = transacao.descricao
            existente.valor = transacao.valor
            transacaoRepository.save(existente)
        } else {
            transacao.dataHoraInclusao = LocalDateTime.now()
            transacaoRepository.save(transacao)

            // Feature: Auto-baixa de contas a pagar pelo valor exato
            if (!transacao.positivo) {
                val contas = contaPagarRepository.findPossivelPagamento(transacao.valor.abs())
                if (contas.size == 1) {
                    val conta = contas[0]
                    conta.pago = true
                    conta.dataPagamento = transacao.data
                    contaPagarRepository.save(conta)
                    logger.info("Conta ${conta.id} baixada automaticamente pela transação ${transacao.id}")
                }
            }
        }
    }

    private fun calcularSaldoDiario(contaBanco: ContaBanco, dataInicio: LocalDateTime): BigDecimal {
        val transacoes = transacaoRepository.findByContaBancoAndPeriodo(
            contaBanco.id, dataInicio, LocalDateTime.now()
        )
        
        var saldoAcumulado = BigDecimal.ZERO
        val saldosPorDia = transacoes.groupBy { it.data.toLocalDate() }
        
        for ((data, transacoesDia) in saldosPorDia) {
            val totalDia = transacoesDia.sumOf { it.valor }
            saldoAcumulado = saldoAcumulado.add(totalDia)
        }
        
        saldoCache["${contaBanco.id}_saldo"] = saldoAcumulado
        return saldoAcumulado
    }
}

data class RetornoHelper(
    var sucesso: Boolean = true,
    var mensagem: String? = null,
    var objeto: Any? = null
)

File: TypeScript/extrato.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { DataSource } from 'typeorm';
import * as amqp from 'amqplib';
import axios from 'axios';
import { TransacaoFinanceiraRepository, ContaBancoRepository, ContaPagarRepository } from './repositories';
import { TransacaoFinanceira, ContaBanco } from './entities';

export interface RetornoHelper {
  sucesso: boolean;
  mensagem?: string;
  objeto?: any;
}

@Injectable()
export class ExtratoService {
  private readonly logger = new Logger(ExtratoService.name);
  private readonly apiUrl: string;
  private readonly apiToken: string;
  private saldoCache: Record<string, number> = {};

  constructor(
    private readonly transacaoRepository: TransacaoFinanceiraRepository,
    private readonly contaBancoRepository: ContaBancoRepository,
    private readonly contaPagarRepository: ContaPagarRepository,
    private readonly dataSource: DataSource,
    private readonly configService: ConfigService,
  ) {
    this.apiUrl = this.configService.get<string>('KAMINO_API_URL');
    this.apiToken = this.configService.get<string>('KAMINO_API_TOKEN');
  }

  async consultarPendentes(companyName: string, idContaBanco: number): Promise<any[] | null> {
    try {
      const url = `${this.apiUrl}/hub/v1/financial/transactions/pendings`;

      const response = await axios.get(url, {
        headers: {
          'Authorization': `Bearer ${this.apiToken}`,
          'X-Company': companyName,
          'X-Account-Id': idContaBanco.toString(),
        },
      });

      if (response.status === 200) {
        return response.data;
      } else {
        this.logger.error(`Erro ao buscar transacoes pendentes: ${response.status}`);
      }
    } catch (error) {
      this.logger.error('Erro ao buscar transacoes pendentes', error);
    }
    return null;
  }

  async getSaldoPendente(companyName: string, idContaBanco: number): Promise<any | null> {
    const cacheKey = `${companyName}_${idContaBanco}`;
    if (this.saldoCache[cacheKey] !== undefined) {
      return { balance: this.saldoCache[cacheKey], cached: true };
    }

    try {
      const url = `${this.apiUrl}/hub/v1/bank/balance/pendings`;

      const response = await axios.get(url, {
        headers: {
          'Authorization': 'Bearer ' + this.apiToken,
          'X-Company': companyName,
        },
      });

      if (response.data?.balance !== undefined) {
        this.saldoCache[cacheKey] = response.data.balance;
      }

      return response.data;
    } catch (error) {
      this.logger.error(`Erro ao buscar saldo pendente: ${error.message}`);
    }
    return null;
  }

  async atualizarContasBancarias(companyToken: string, forcar: boolean = false): Promise<RetornoHelper> {
    let reagendar = true;
    try {
      const contasBanco = await this.contaBancoRepository.findContasIntegradas();

      if (contasBanco.length === 0) {
        reagendar = false;
        return { sucesso: true };
      }

      this.logger.log(`Atualizando ${contasBanco.length} contas bancarias`);

      for (const contaBanco of contasBanco) {
        await this.sincronizarConta(contaBanco, companyToken, forcar);
      }

      return { sucesso: true };
    } catch (error) {
      this.agendarSincronizacao(companyToken);
      this.logger.error('Erro ao atualizar extrato contas bancarias', error);
      return { sucesso: false, mensagem: 'Nao foi possivel atualizar as contas bancarias' };
    } finally {
      if (reagendar) {
        this.agendarSincronizacao(companyToken);
      }
    }
  }

  async agendarSincronizacao(companyToken: string): Promise<void> {
    try {
      const connection = await amqp.connect('amqp://localhost');
      const channel = await connection.createChannel();

      await channel.assertQueue('broker-queue', { durable: false });

      const message = {
        companyToken,
        metodo: 'POST',
        endpoint: 'api/banking/kamino/transactions/reload/hook',
        dataHoraProgramacao: new Date(Date.now() + 5 * 60 * 1000).toISOString(),
      };

      channel.sendToQueue('broker-queue', Buffer.from(JSON.stringify(message)));
    } catch (error) {
      this.logger.error('Erro ao agendar broker de atualizacao', error);
    }
  }

  async sincronizarConta(contaBanco: ContaBanco, companyToken: string, forcar: boolean = false): Promise<RetornoHelper> {
    this.logger.log(`Iniciando atualizacao da conta bancaria: ${contaBanco.id}`);

    const queryRunner = this.dataSource.createQueryRunner();
    await queryRunner.connect();
    await queryRunner.startTransaction();

    try {
      const ultimaTransacao = await this.transacaoRepository.findUltimaTransacao(contaBanco.id);

      let dataInicio = ultimaTransacao?.data ?? new Date('2023-01-01');
      dataInicio = new Date(dataInicio.getTime() - 24 * 60 * 60 * 1000);
      if (forcar) dataInicio = new Date('2023-01-01');

      const transacoesExternas = await this.buscarTransacoesAPI(contaBanco.id, dataInicio, new Date());
      this.logger.log(`Transacoes obtidas da API: ${transacoesExternas.length}`);

      for (const transacaoExterna of transacoesExternas) {
        const transacao = this.converterParaEntidade(transacaoExterna, contaBanco);
        await this.salvarTransacao(transacao);
      }

      await this.calcularSaldoDiario(contaBanco, dataInicio);

      await queryRunner.commitTransaction();
      return { sucesso: true };
    } catch (error) {
      await queryRunner.rollbackTransaction();
      throw error;
    } finally {
      await queryRunner.release();
    }
  }

  private async buscarTransacoesAPI(idContaBanco: number, dataInicio: Date, dataFim: Date): Promise<any[]> {
    const formatDate = (d: Date) => d.toISOString().split('T')[0];
    const url = `${this.apiUrl}/hub/v1/financial/transactions?startDate=${formatDate(dataInicio)}&finalDate=${formatDate(dataFim)}&accountId=${idContaBanco}`;

    try {
      const response = await axios.get(url, {
        headers: {
          'Authorization': `Bearer ${this.apiToken}`,
        },
      });

      return response.data?.items ?? [];
    } catch (error) {
      this.logger.error('Erro ao buscar transacoes da API', error);
      return [];
    }
  }

  private converterParaEntidade(dados: any, contaBanco: ContaBanco): TransacaoFinanceira {
    const transacao = new TransacaoFinanceira();
    transacao.codigoNoBanco = dados.id;
    transacao.descricao = dados.description;
    transacao.valor = Number(dados.amount);
    transacao.data = new Date(dados.date);
    transacao.idContaBanco = contaBanco.id;
    transacao.idTransacaoExterna = dados.externalId;
    return transacao;
  }

  async salvarTransacao(transacao: TransacaoFinanceira): Promise<void> {
    const existente = await this.transacaoRepository.findByCodigoNoBanco(transacao.codigoNoBanco ?? '');

    if (existente) {
      existente.descricao = transacao.descricao;
      existente.valor = transacao.valor;
      await this.transacaoRepository.save(existente);
    } else {
      transacao.dataHoraInclusao = new Date();
      await this.transacaoRepository.save(transacao);

      // Feature: Auto-baixa de contas a pagar pelo valor exato
      if (!transacao.positivo) {
        const contas = await this.contaPagarRepository.findPossivelPagamento(Math.abs(transacao.valor));
        if (contas.length === 1) {
          const conta = contas[0];
          conta.pago = true;
          conta.dataPagamento = transacao.data;
          await this.contaPagarRepository.save(conta);
          this.logger.log(`Conta ${conta.id} baixada automaticamente pela transacao ${transacao.id}`);
        }
      }
    }
  }

  private async calcularSaldoDiario(contaBanco: ContaBanco, dataInicio: Date): Promise<number> {
    const transacoes = await this.transacaoRepository.findByContaBancoAndPeriodo(
      contaBanco.id,
      dataInicio,
      new Date(),
    );

    const saldosPorDia: Record<string, number> = {};
    for (const transacao of transacoes) {
      const dia = transacao.data.toISOString().split('T')[0];
      saldosPorDia[dia] = (saldosPorDia[dia] ?? 0) + transacao.valor;
    }

    let saldoAcumulado = 0;
    for (const dia of Object.keys(saldosPorDia).sort()) {
      saldoAcumulado += saldosPorDia[dia];
    }

    this.saldoCache[`${contaBanco.id}_saldo`] = saldoAcumulado;
    return saldoAcumulado;
  }
}
